File: IndividualAssignment1/Code/C++/Matrix.cpp
#include "Matrix.h"
#include <random>
#include <algorithm>

Matrix random_matrix(int n, unsigned seed) {
    std::mt19937_64 rng(seed);
    std::uniform_real_distribution<double> dist(0.0, 1.0);
    Matrix M(n, std::vector<double>(n));
    for (int i = 0; i < n; i++)
        for (int j = 0; j < n; j++)
            M[i][j] = dist(rng);
    return M;
}

Matrix zero_matrix(int n) {
    return Matrix(n, std::vector<double>(n, 0.0));
}

Matrix naive_multiply(const Matrix &A, const Matrix &B) {
    int n = (int)A.size();
    Matrix C = zero_matrix(n);
    for (int i = 0; i < n; i++) {
        for (int k = 0; k < n; k++) {
            double aik = A[i][k];
            for (int j = 0; j < n; j++) {
                C[i][j] += aik * B[k][j];
            }
        }
    }
    return C;
}

Matrix blocked_multiply(const Matrix &A, const Matrix &B, int block) {
    int n = (int)A.size();
    Matrix C = zero_matrix(n);
    for (int ii = 0; ii < n; ii += block) {
        for (int kk = 0; kk < n; kk += block) {
            for (int jj = 0; jj < n; jj += block) {
                int iMax = std::min(ii + block, n);
                int kMax = std::min(kk + block, n);
                int jMax = std::min(jj + block, n);
                for (int i = ii; i < iMax; i++) {
                    for (int k = kk; k < kMax; k++) {
                        double aik = A[i][k];
                        for (int j = jj; j < jMax; j++) {
                            C[i][j] += aik * B[k][j];
                        }
                    }
                }
            }
        }
    }
    return C;
}

File: IndividualAssignment1/Code/Python/Matrix.py

"""
matrix multiplication implementations.
Production code only (no CLI / benchmarking code here).
"""

from typing import List, Callable, Optional
import random

matrix = List[List[float]]

def make_random_matrix(n: int, seed: Optional[int]=None) -> matrix:
    if seed is not None:
        random.seed(seed)
    return [[random.random() for _ in range(n)] for _ in range(n)]

def zero_matrix(n: int) -> matrix:
    return [[0.0 for _ in range(n)] for _ in range(n)]

def multiply_naive(A: matrix, B: matrix) -> matrix:
    n = len(A)
    C = zero_matrix(n)
    for i in range(n):
        Ai = A[i]
        Ci = C[i]
        for k in range(n):
            Aik = Ai[k]
            Bk = B[k]
            for j in range(n):
                Ci[j] += Aik * Bk[j]
    return C

def multiply_blocked(A: matrix, B: matrix, block: int=32) -> matrix:
    n = len(A)
    C = zero_matrix(n)
    for ii in range(0, n, block):
        for kk in range(0, n, block):
            for jj in range(0, n, block):
                i_max = min(ii+block, n)
                k_max = min(kk+block, n)
                j_max = min(jj+block, n)
                for i in range(ii, i_max):
                    for k in range(kk, k_max):
                        Aik = A[i][k]
                        rowC = C[i]
                        rowB = B[k]
                        for j in range(jj, j_max):
                            rowC[j] += Aik * rowB[j]
    return C

def multiply_numpy(A, B):
    try:
        import numpy as np
    except ImportError:
        raise RuntimeError("numpy not installed")
    return (np.dot(A, B)).tolist()
